### ISSUE TITLE
Passing OAuth access_token to upstream via Authorisation header (rather than X-Forwarded-Access-Token)

### ISSUE THREAD

**Issue by k-wall** [openshift/oauth-proxy#107]
In our use-case, we want to use oauth-proxy to protect a user-facing application (HTML/JS single page app).

oauth-proxy will be used to allow the user to authenticate against the OpenShift OAuth server, serve the application's static content, and provide a proxy to an API server.   The user's interactions with the application will cause API calls to be made.   The application won't be in possession of the access_token, so I wanted to rely on oauth-proxy to argument the authenticated request with the authentication user's access token as it traversed the oauth-proxy.

I have found `--pass-access-token=true` which does almost what I want, however the API server in question does not understand X-Forwarded-Access-Token.  I desire a version of the `--pass-access-token` option that would pass OAuth access_token of the authenticated user to upstream via Authorisation header formatted in the bearer format..

I can't change the API server to understand X-Forwarded-Access-Token (the API server in question is the Kubernetes API server itself).

Is there another way to use oauth-proxy that would achieve the same ends?  Alternatively, I would be happy to submit a PR.

Thanks in advance.

**Comment by InfoSec812** [comment]
I'm in the same situation. I thought that `--pass-user-bearer-token=true` would solve this, but I am getting odd results. For example, my config looks like:

```
        containers:
        - name: oauth-proxy
          image: openshift/oauth-proxy:latest
          imagePullPolicy: IfNotPresent
          ports:
          - containerPort: ${{PROXY_PORT}}
            protocol: TCP
          args:
          - "--https-address="
          - "--http-address=:${PROXY_PORT}"
          - "--provider=openshift"
          - "--openshift-service-account=proxy"
          - "--upstream=https://openshift.default.svc:443/oapi"
          - "--upstream=https://openshift.default.svc:443/api"
          - "--upstream=http://localhost:${APP_PORT}"
          - '--cookie-name=OCP_TOKEN'
          - '--cookie-expire=24h0m0s'
          - '--cookie-refresh=8h0m0s'
          - "--cookie-domain=.rht-labs.com"
          - "--cookie-name=OCP_TOKEN"
          - "--cookie-secret=${COOKIE_SECRET}"
          - "--pass-user-bearer-token=true"
```

But when my front-end makes a request through the proxy for `/oapi/v1` I get an NGinx 404 page from the UI server instead of a response from the OpenShift API.

**Comment by InfoSec812** [comment]
@enj or @stlaz Any insights on this?

**Comment by InfoSec812** [comment]
Hmmm... I turned on request logging and the proxy is most definitely not forwarding the request to the alternate upstream:

```
10.131.0.1 - josphill@cluster.local [26/Mar/2019:14:24:13 +0000] oauth-test-deven-test.apps.s11.core.rht-labs.com GET localhost:8080 "/oapi/v1" HTTP/1.1 "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36" 404 571 0.001
```

**Comment by k-wall** [comment]
@InfoSec812  Maybe our use cases are different.  If I understand the semantics of `--pass-user-bearer-token` correctly, it would take the OAuth access token received in the request from the _client_ and forward it to the upstream contained in a X-Forwarded-Access-Token header.  That's not what I want.  In my use-case the client (which is a single page web-app) does not see the access token at all.  I want the access_token from the authenticated session held by oauth-proxy to be forwarded with the client's API requests as an `Authorisation: bearer ...` header.

In other words, I desire is a flag like https://github.com/pusher/oauth2_proxy's -pass-authorization-header  flag:

`  -pass-authorization-header: pass OIDC IDToken to upstream via Authorization Bearer header
`

except passing the OAuth Access Token rather than  OIDC IDToken .

Like I say above, I'd be happy to put up a PR.

**Comment by InfoSec812** [comment]
@k-wall That's exactly what I want... I may perhaps be misunderstanding the use of `--pass-user-bearer-token`. 

I did some further digging, and I found the issue with my paths not being mapped correctly and it has to do with Golang's `httpServMux` requiring that the paths registered end with a trailing `/` (slash). Once I added the trailing slash, it is now trying to forward my request. This, however, lead me to a new problem and I will open a separate issue for that.
